### PR TITLE
WIP:Issue #343: CircuitBreaker should only allow a certain number of test…

### PR DESCRIPTION
**Pull request by RobWin**
… requests in HALF_OPEN and reject calls when maximum number is reached. The number of requests is equal to ringBufferSizeInHalfOpenState.

### CODE REVIEW

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerState.java [hunk 1]
```diff
@@ -32,7 +32,9 @@
         this.stateMachine = stateMachine;
     }
 
-    abstract boolean isCallPermitted();
+    abstract boolean obtainPermission();
```
Comment (leonard84) [inline resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerState.java]: Having done some C# programming I would actually switch how you implemented them, `tryObtainPermission()` should return `boolean` and `obtainPermission()` should throw an exception.

See https://stackoverflow.com/questions/467613/parse-v-tryparse

Comment (RobWin) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerState.java]: Yes makes sense.

Comment (RobWin) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerState.java]: Be aware that the PR is still work in progress. But feel free to review  it

**Comment by Romeh** [comment]
@RobWin  we do not need to follow the same naming for Fixed thread pool bulkhead or this will be done on another PR ?

**Comment by RobWin** [comment]
Same naming, should be done in this PR

**Pull request merged**